geode-membership/src/main/java/org/apache/geode/distributed/internal/membership/gms/GMSMembership.java

```diff
@@ -1231,23 +1197,16 @@ public void shutdownMessageReceived(ID id, String reason) {
     if (logger.isDebugEnabled()) {
       logger.debug("Membership: recording shutdown status of {}", id);
     }
-    synchronized (this.shutdownMembers) {
-      this.shutdownMembers.put(id, id);
-      services.getHealthMonitor()
-          .memberShutdown(id, reason);
-      services.getJoinLeave().memberShutdown(id, reason);
-    }
+    this.shutdownMembers.add(id);
+    services.getJoinLeave().memberShutdown(id, reason);
   }
 
   @Override
   public Set<ID> getMembersNotShuttingDown() {
-    latestViewReadLock.lock();
-    try {
-      return latestView.getMembers().stream().filter(id -> !shutdownMembers.containsKey(id))
+    synchronized (shutdownMembers) {
```

Comment:
Only a single thread can read this set of members now. Previously all threads could read concurrently.

Reply:
Previously reading threads waited for the `latestViewReadLock`.

But also there was previously a latent bug here, to wit, puts to the non-concurrent collection `shutdownMembers` were protected (via synchronization on the collection) whereas the read access to that collection by the `filter()` loop here was not protected via synchronization.

I think synchronization here is a valid fix for that problem.

Reply:
Yes, as Bill said, `shutdownMembers` was a non-concurrent collection, and the read access to it wasn't synched. Although `shutdownMembers` is a `synchronizedSet` now, synchronization is still needed around `stream()` operation.

Reply:
You are saying that currently the shutdownMembers collection can be mutated without acquiring the write lock? If that is the case wouldn't the more appropriate fix be to acquire the write lock before mutation rather than adding a serializing synchronization block here?

Reply:
Or give `shutdownMembers` the same immutability treatment that `lastView` has.

Reply:
Doing so also allows you to make the "not shutting down" members list cacheable rather than creating a new list every time the method is called.

Reply:
Since this is a potentially long linked list I don't think copy-on-write is a good pattern to use for shutdownMembers.
This method is invoked on every message-send, so maybe using a synchronized collection isn't a great idea and we should use the read/write lock instead.

Reply:
If the list isn't updated frequently even a long list with copy on write would be better less intrusive than frequently grabbing a read lock. If this method is called for every p2p message we certainly can't have it synchronizing, it should do the least amount of blocking as possible.

Reply:
Wait a second, we are creating a new set of members not shutting down for every p2p message sent?

Reply:
@pivotal-jbarrett yes, we've always created a new collection like this for the reply processors.  This code now seems to create two (if you look at its primary use in ClusterDistributionManager) so that could be made more efficient.

Reply:
If the `shutdownMembers` list isn't changing frequently, like every ms, then repeatedly recreating the same set is just adding pressure to the GC and CPU time all while under a synchronized lock for no real gain. I assume this `shutdownMembers` list is updated with the member id of the member that has said it is or has gone down, which isn't a frequent event in production. Seems like we should be optimizing this method to be returning a rather static list of members known not to be shutting down even if that means the infrequent setting of a shutting down member takes a bit more time to handle, by copying lists and invalidating or updating the not shutdown servers list. 

At the very least the addition of this synchronized block is going to have a noticeable impact on the p2p messaging at scale since all messages will be serialized through this lock. If we deem refactoring out the list creation is out of scope for this PR then using a read/write lock here should be in scope as a first step.

Reply:
@pivotal-jbarrett We'll consider creating another ticket to optimize the method, as you said. For now, I replaced the synchronized blocks with read/write locks.

Reply:
Awesome! Thanks!